src/main/java/com/fasterxml/jackson/core/JsonGenerator.java

```diff
@@ -2482,22 +2482,18 @@ public void copyCurrentEvent(JsonParser p) throws IOException
             NumberType n = p.getNumberType();
             if (n == NumberType.INT) {
                 writeNumber(p.getIntValue());
-            } else if (n == NumberType.BIG_INTEGER) {
-                writeNumber(p.getBigIntegerValue());
             } else {
-                writeNumber(p.getLongValue());
+                writeNumber(p.getBigIntegerValue());
             }
             break;
         }
         case ID_NUMBER_FLOAT:
         {
             NumberType n = p.getNumberType();
-            if (n == NumberType.BIG_DECIMAL) {
-                writeNumber(p.getDecimalValue());
-            } else if (n == NumberType.FLOAT) {
+            if (n == NumberType.FLOAT) {
                 writeNumber(p.getFloatValue());
             } else {
-                writeNumber(p.getDoubleValue());
```

Comment:
This is the fix, but only need if `StreamReadCapability.EXACT_FLOATS` is false. The reason we want to check that is to prevent unnecessary coercion when there is true 64-bit `double` under the hood.

So I think something like 

    else if (n == NumberType.DOUBLE && p.getReadCapabilities().isEnabled(StreamReadCapability.EXACT_FLOATS)) {
       writeNumber(p.getDoubleValue());
    } else {
       writeNumber(p.getDecimalValue());

might work. Although, maybe even better... `parser.getNumberValueExact()`?

Reply:
Ah. Whether `getNumberValueExact()` would work or not, it's inconvenient...

Reply:
I changed the code to use `getNumberValueExact()`

Reply:
Hmmh. Yeah there's something wrong there (diff?) -- there's combination of Int and FP handling.

But I also realized that use of `getNumberValueExact()` might not be the best way unfortunately, returning `Number`, requiring casting. I could try to see if what I suggested above works better.